### DISCUSSION TITLE
Optuna without an objective function?

### DISCUSSION THREAD

**Discussion by DeLaRiva** [optuna/optuna#3604] [Q&A] [answered]
Hi,

I would like to use Optuna but I lack an objective function. My own objective is the following:

There are multiple devices which operate at individual parameter settings and I want to find their optimal values w.r.t. to maximize a target variable which depends on those parameters.
To be more precise: The devices are x-ray tubes for medical applications which have parameters like current, voltage, angle and so on and the target variable is the deposited energy aka dose. The dose depends on current, voltage.. which vary for each tube and there is no function available to determine the dose analytically.

My question now is: Is it possible to use Optuna to find the parameters which maximize the dose?
From all examples shown on the website I couldn't find certain examples or the like which realize this or something similar but I would like to make sure I'm not missing something.
Thanks a lot in advance!

**Comment by Rjdrenth** [comment]
I am a new user of Optuna, but I think I can answer your question anyway, as your question does not really seem related to Optuna. 

You have 2 challenges:
1. The tubes behave differently 
2. You don't have a mathematical function available to determine their doses.

The first problem can be solved by optimising the parameters for each tube individually. Not ideal, but that would be the way to go.

The second problem could be solved in 2 ways: 
- Hook up the tubes to your optimisation code, so that a new parameter combination can be tested on the actual hardware and feed the measured dose back to the optimisation system. This is the best way, as you will have reliable results and don't have to perform manual work. If you cant do this in an automated fashion, you can of course operate the hardware manually and feed back the results to your optimisation code. 
- If the first solution is not possible or practical, you could perhaps perform (manual?) experiments where you try different settings and determine the dose. Then you could feed that into any approximator (could be simple lineair regression or another statistical / ML model), which will estimate the tube's dose given a certain combination of parameters. Then you could use such a trained model with Optuna to determine the optimal set of parameters. 
You do need to realise that this method might not be reliable if the approximator is unable to properly model the behaviour of the tubes, but it can give you a 'good enough' approximation (but what is good enough in a medical context?..) 

Good luck!

**Comment by DeLaRiva** [comment]
Hi Rjdrenth,

thanks a lot for your detailed input! I confirm in all aspects you araised and I'm quite glad that I came towards the same results.

Fortunately, I am able to move on with the second problem in the first manner: This means, I am able to give the device parameters which tests these and feeds back the yielded dose to Optuna.

So far, my "only problem is, that I don't know how to use Optuna (in this context) without an objective function?
Or in other words, how does Optuna know what to optimize when there is no function provided?

Could you please share me some thoughts or even instructions? In any case, thanks a lot in advance!

**Comment by Rjdrenth** [comment, marked as answer]
Hi, no problem. 

You would have to write your own objective function where you indicate which parameters you need to optimise, perform the experiment on your tube and return the resulting dose. The following code should provide you with a starting point for what you need to implement:

```python
import optuna

# Objective function that takes three arguments.
def objective(trial, tube_name):

    # TODO Fill in proper min / max values and add all relevant parameters
    MIN_CURRENT = 0
    MAX_CURRENT = 1
    current = trial.suggest_float("current", MIN_CURRENT, MAX_CURRENT)

    MIN_VOLTAGE = 0
    MAX_VOLTAGE = 1
    voltage = trial.suggest_float("voltage", MIN_VOLTAGE, MAX_VOLTAGE)

    MIN_ANGLE = 0
    MAX_ANGLE = 180
    angle = trial.suggest_int("angle", MIN_ANGLE, MAX_ANGLE)

    # TODO implement method "perform_experiment_with_tube", which should:
    #  - connect to the tube and submit the selected parameters (via API request, TCP connection, websocket, whatever)
    #  - wait for the resulting dose as reported by your system and return this value
    dose = perform_experiment_with_tube(tube_name, current, voltage, angle)

    return dose


if __name__ == "__main__":

    # TODO Define the proper names with which you will connect to your tubes
    tube_names = [
        "tube_1",
        "tube_2",
        "tube_3",
        "tube_alpha"
        "tube_whatever"
    ]

    # TODO Determine logical number of trials
    n_trials = 50
  
    # Perform the optimisation for all tubes, 50 trials are performed per tube
    for tube_name in tube_names:
        study_name = f"tube_{tube_name}_optimisation"
        storage_name = f"sqlite:///{study_name}.sqlite3"  # .format(str(paths.MODELS_PATH / study_name))

        # Create and execute the study
        study = optuna.create_study(study_name=study_name, storage=storage_name, load_if_exists=True, direction="maximize")
        study.optimize(lambda trial: objective(trial, tube_name), n_trials=n_trials, show_progress_bar=True)
```

**Comment by DeLaRiva** [comment]
wow, nice, thank you very much!

To understand it: You didn't set up an objective function, right? With function I mean an analytical function.

Actually, through perform_experiment_with_tube() I would write the parameters to the tubes and would receive the gained dose.
And I don't need to clarify the relationship between the parameters and the dose?

I'm curious how it will work but I'll probably need some time to get a proper connection for I/O.

Again, thank you!

**Reply by Rjdrenth** [reply]
No problem! 

I indeed did not set up any analytical function, as it is not necessary to do so as you can make a connection with the tubes directly.
It also is not necessary to clarify the relationship between parameters and the dose; Optuna will try to figure that out by itself.

Good luck!

**Reply by DeLaRiva** [reply]
Again, thanks a lot!

**Comment by DeLaRiva** [comment]
Hi,

it is probably a programming mistake on my side but in case it has some deeper meaning, I just wanted to ask you first:

I modified your code slightly

```python
def objective(trial, tube_name):

    import input
    import sys

    # TODO Fill in proper min / max values and add all relevant parameters
    MIN_CURRENT = 0
    MAX_CURRENT = 1
    current = trial.suggest_float("current", MIN_CURRENT, MAX_CURRENT)

    MIN_VOLTAGE = 0
    MAX_VOLTAGE = 1
    voltage = trial.suggest_float("voltage", MIN_VOLTAGE, MAX_VOLTAGE)

    MIN_ANGLE = 0
    MAX_ANGLE = 180
    angle = trial.suggest_int("angle", MIN_ANGLE, MAX_ANGLE)

    # print parameters
    print("Tube_name", {tube_name})
    print("Current", {current})
    print("Voltage ", {voltage )
    print("Angle", {angle})
    
    # receive dose which will be entered by keyboard
    dose = input("Please enter the yielded dose: ")
    print(f'You entered {dose}')

    return dose
```

I thought, I could investigate the potential of Optuna already before connecting several devices with each other simply by entering the parameters in the forward-direction and the dose in the feedback-direction.

The only issue I face now is, that I can only enter a dose between 0-9, every more digits, like 10 or 90234 are forbidden and lead e.g. to following responce: 

> Trial 0 failed, because the number of the values 2 did not match the number of the objectives 1

 when I enter "10" per keyboard as dose. I'm not sure whether to focus on a programming issue or on a conceptual issue (wrt to Optuna).

Do you mind shedding some light on me?
Thanks a lot in advance!

**Comment by Rjdrenth** [comment]
Optuna expects an integer or a float to be returned. If you use the
`input(..)` method, a string will be returned. After doing `dose =
input(..)` you will need to do `dose = float(dose)`.

You can check the type of a variable with `print(type(dose))` to find out
more about it.

**Comment by kinthaiofficial** [comment]
Optuna without an explicit objective function is possible through a few different patterns, depending on what you're trying to avoid:

**Pattern 1: Define objective inline (reduce boilerplate)**
If the concern is repetitive objective function definitions:
```python
import optuna

def optimize_model(X_train, y_train, X_val, y_val, n_trials=100):
    def objective(trial):
        # Parameters defined inline
        params = {
            "n_estimators": trial.suggest_int("n_estimators", 50, 300),
            "max_depth": trial.suggest_int("max_depth", 3, 10),
            "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3),
        }
        model = xgb.XGBClassifier(**params)
        model.fit(X_train, y_train)
        return accuracy_score(y_val, model.predict(X_val))
    
    study = optuna.create_study(direction="maximize")
    study.optimize(objective, n_trials=n_trials)
    return study.best_params
```

**Pattern 2: `optuna.terminator` for external metric injection**
If the metric comes from an external system (a training run that logs metrics), use the terminator callback pattern to let external processes report back.

**Pattern 3: Manual trial management**
```python
study = optuna.create_study()
trial = study.ask()  # Gets a trial without running objective
params = {
    "lr": trial.suggest_float("lr", 1e-5, 1e-1)
}
# Run training externally with these params
accuracy = run_training(params)
study.tell(trial, accuracy)  # Report result back
```

This is the real "Optuna without objective function" pattern — decouple parameter suggestion from metric reporting. Useful for: external training jobs, distributed training across machines, manual evaluation.

What's the specific use case driving this?
